torchao/prototype/mx_formats/mx_tensor.py

```diff
@@ -0,0 +1,411 @@
+    # if the underlying data is transposed, convert to row major before
+    # unpacking and unscaling
+    if is_transposed:
+        data_lp = data_lp.t()
+        assert data_lp.is_contiguous()
+        orig_shape = (orig_shape[1], orig_shape[0])
+
+    if elem_dtype in (torch.float8_e4m3fn, torch.float8_e5m2):
+        data_hp = data_lp.to(target_dtype)
+    elif elem_dtype == DTYPE_FP6_E2M3:
+        data_hp = f6_e2m3_unpacked_to_f32(data_lp)
+        data_hp = data_hp.to(target_dtype)
+    elif elem_dtype == DTYPE_FP6_E3M2:
+        data_hp = f6_e3m2_unpacked_to_f32(data_lp)
+        data_hp = data_hp.to(target_dtype)
+    elif elem_dtype == DTYPE_FP4:
+        if config.use_fp4_custom_triton_dequant_kernel:
+            data_hp_rescaled = triton_f4_to_scaled_bf16(
+                data_lp,
+                scale_e8m0,
+                block_size,
+            )
+            if is_transposed:
+                data_hp_rescaled = data_hp_rescaled.t()
+            return data_hp_rescaled.to(target_dtype)
+        else:
+            # fp4
+            f4_unpacked = unpack_uint4(data_lp)
+            # for now we only have a cast to f32
+            # TODO(future PR): add cast directly to bf16
+            f32 = f4_unpacked_to_f32(f4_unpacked)
+            data_hp = f32.to(target_dtype)
+            # manually adjust shape to account for the unpacking
+            # TODO(future PR): clean up the shape code and remove the hack
+            # below
+        orig_shape = (*orig_shape[:-1], orig_shape[-1] * 2)
+    else:
+        raise AssertionError("unsupported")
+
+    data_hp = data_hp.reshape(-1, block_size)
+    s_fp = get_fp_scale(scale_e8m0).reshape(-1, 1).to(target_dtype)
+    data_hp = data_hp * s_fp
+    data_hp = data_hp.reshape(orig_shape)
+
+    # if we converted to row-major before unscaling convert back
+    if is_transposed:
+        data_hp = data_hp.t()
+
+    return data_hp
+
+
+def tensor_size_hp_to_fp4x2(orig_size, is_contiguous):
+    new_size = orig_size
+    if is_contiguous:
+        new_size = [*list(new_size[:-1]), new_size[-1] // 2]
+    else:
+        new_size = [new_size[0] // 2, *list(new_size[1:])]
+    return new_size
+
+
+def tensor_size_fp4x2_to_hp(orig_size, is_contiguous):
+    new_size = orig_size
+    if is_contiguous:
+        new_size = [*list(new_size[:-1]), new_size[-1] * 2]
+    else:
+        new_size = [new_size[0] * 2, *list(new_size[1:])]
+    return new_size
+
+
+@torch._dynamo.allow_in_graph
+class ToMXConstrFunc(torch.autograd.Function):
+    """
+    Differentiable cast to MX, no-op in backward
+    """
+
+    @staticmethod
+    def forward(ctx, data_hp, elem_dtype, block_size):
+        scale_e8m0_biased, data_lp = to_mx(data_hp, elem_dtype, block_size)
+        return MXTensor(
+            scale_e8m0_biased, data_lp, elem_dtype, block_size, data_hp.dtype
+        )
+
+    @staticmethod
+    def backward(ctx, g):
+        return g, None, None
+
+
+@torch._dynamo.allow_in_graph
```

Comment:
There's a public API `torch.compiler.allow_in_graph` - also curious why this was needed

Reply:
this is necessary for compile to fully support training, and this line is copy-pasta from float8_experimental, ideally while these two products are in different codebases I'm hoping for these kind of issues to get fixed in float8_experimental first and be copied here.  Once we unify it will be easier.